ArduinoJson v7 Migration & Developer Experience Improvements

HadyBazzi opened the pull request:
## ArduinoJson v7 Migration & Developer Experience Improvements

### Summary

This PR modernizes the library with ArduinoJson v7 compatibility, adds SSL/TLS support for production security, and significantly improves the developer experience by enabling configuration overrides without editing library source files.

### Changes

#### New Features

- **Configuration overrides** — Users can now configure the library via `#define` statements before including the library (Arduino IDE) or via `build_flags` (PlatformIO), eliminating the need to edit `ESP32_AI_Connect_config.h`
- **SSL/TLS support** — New `setRootCA()` and `getRootCA()` methods allow enabling secure connections with certificate verification (disabled by default for backwards compatibility and ease of use)

#### Refactoring

- **ArduinoJson v7 migration** — Updated all deprecated JSON handling to use v7 API:
  - `DynamicJsonDocument` / `StaticJsonDocument` → `JsonDocument`
  - `containsKey()` → `!field.isNull()`
  - `createNestedArray()` / `createNestedObject()` → `.to<JsonArray>()` / `.to<JsonObject>()`

#### Documentation

- Updated all examples and user guides to reflect the new configuration pattern
- Added SSL documentation to README and User Guides along with example code
- Updated `keywords.txt` with new methods

### New Configuration Pattern

```cpp
// Arduino IDE - define BEFORE including the library
#define DISABLE_AI_API_GEMINI
#define AI_API_REQ_JSON_DOC_SIZE 8192
#include <ESP32_AI_Connect.h>
```

```ini
# PlatformIO - add to platformio.ini
build_flags =
    -DDISABLE_AI_API_GEMINI
    -DAI_API_REQ_JSON_DOC_SIZE=8192
```

### New SSL/TLS Feature

Enable secure connections with certificate verification for production deployments:

```cpp
const char* root_ca = R"(
-----BEGIN CERTIFICATE-----
XXXXXXXXXX
XXXXXXXXXX
XXXXXXXXXX
-----END CERTIFICATE-----
)";

// Enable secure mode
aiClient.setRootCA(root_ca);

// Check security status
if (aiClient.getRootCA() != nullptr) {
    Serial.println("Secure mode enabled");
}

// Disable secure mode (back to insecure)
aiClient.setRootCA(nullptr);
```

HadyBazzi committed:
feat(config): allow config overrides without editing source files

HadyBazzi committed:
feat(ssl): add root ca certificate configuration

HadyBazzi committed:
docs: modernize configuration override pattern and add SSL documentation

HadyBazzi committed:
refactor(json): migrate to arduinojson v7 api

HadyBazzi commented:
Hello,

I just wanted to follow up on this.

Thanks,
Hady

AvantMaker commented:
> Hello,
> 
> I just wanted to follow up on this.
> 
> Thanks, Hady

Hi @HadyBazzi, thanks a lot for this PR and for the ArduinoJson v7 migration work!

I recently pushed some updates to the repo and tagged the latest release (v0.5.16) after this PR was opened. To avoid breaking existing sketches and examples, could you please:

Rebase or merge the latest main into your branch

Re-run your tests/examples against the current main (and the 0.5.16 release layout)

Confirm in a comment which examples or use cases you tested, and on which board(s)

If you notice any conflicts or behavior changes with the new release structure, feel free to adjust the PR accordingly or point them out so we can discuss the best approach.

Thanks again for contributing and helping improve ESP32_AI_Connect!

HadyBazzi committed:
feat: merge upstream new AI LLM xAI Grok supported and automatic retry feature

HadyBazzi commented:
Hey!

I just merged the latest main (v0.5.16) into my branch and resolved the conflicts.

I made a few tweaks to the auto-retry feature while merging. I changed it to be enabled by default to match how the other features (tool calls, streaming, etc.) work to make it more consistent with the new config format. Also added `#ifndef` guards around all the config values so users can override them with `#define` before including the library (again more consistent with new config format).

Found one small bug too: when `_httpClient.begin()` fails, the response code was staying at 0, which meant connection failures wouldn't retry. Fixed it by setting the code to -1 in that case.

Haven't had a chance to test on actual hardware yet but I will update once I've run through the examples on a board.

HadyBazzi committed:
refactor(json): complete arduinojson v7 migration for grok, examples, and docs

HadyBazzi committed:
fix(examples): improve platformio compatibility and consistency

HadyBazzi commented:
Alright I just made a simple fix and finished refactoring for the ArduinoJson v7 migration. The `resilience_retry` and `secure_connection_demo` examples both worked perfectly on my ESP32 Dev Module (ESP-WROOM-32D).

Should be all set to merge now!

HadyBazzi commented:
Hello,

I just wanted to follow up on this. Are there any other fixes to be made?
